### PR TITLE
feat(DASH): Add MPD Patch support

### PR DESCRIPTION
**Pull request by dave-nicholas**
Closes https://github.com/shaka-project/shaka-player/issues/2228

### CODE REVIEW

**Comment by joeyparrish** [comment]
Thank you for contributing!

I don't understand the patch adapter, though.  I worry that it will not be representative of actual patch streams.  I also see it being applied in the demo to a VOD manifest, which should have no need of patching.  Patching is only useful for updates to live streams, correct?

**Comment by dave-nicholas** [comment]
> Thank you for contributing!
> 
> I don't understand the patch adapter, though. I worry that it will not be representative of actual patch streams. I also see it being applied in the demo to a VOD manifest, which should have no need of patching. Patching is only useful for updates to live streams, correct?

Hi @joeyparrish 

I agree with you that patch manifest adapter is not ideal, but I wanted to give the ability to test the feature in the demo app as open patch compliant streams are not easy to come by (at least I haven't seen any 😆 ). The reason I chose the ad enabled VOD is because of its behaviour of presenting new segments and periods regularly like a live stream.

I can remove remove the manifest adapter if you wish.

**Comment by dave-nicholas** [comment]
Hi @joeyparrish I have removed the patch adapter.
Please let me know if there any concerns or changes in approach that you would like me to make.
We will be taking this change into production shortly.

**Comment by Iragne** [comment]
Happy to support in this kink of new feature.
Let me know if you need help

File: lib/dash/dash_parser.js [hunk 1]
```diff
@@ -497,6 +555,8 @@ shaka.dash.DashParser = class {
 
     // These steps are not done on manifest update.
     if (!this.manifest_) {
+      await this.periodCombiner_.combinePeriods(periods, context.dynamic);
```
Comment (joeyparrish) [inline lib/dash/dash_parser.js:558]: This was just run on line 542.  Why do it again now?

File: lib/dash/dash_parser.js [hunk 2]
```diff
@@ -528,25 +588,240 @@ shaka.dash.DashParser = class {
...
+              context.representation.id) {
+          representationIds.push(context.representation.id);
+        }
+      }
     }
 
-    // Add text streams to correspond to closed captions.  This happens right
-    // after period combining, while we still have a direct reference, so that
-    // any new streams will appear in the period combiner.
-    this.playerInterface_.makeTextStreamsForClosedCaptions(this.manifest_);
```
Comment (joeyparrish) [inline lib/dash/dash_parser.js:549]: What replaces this call you deleted?

File: lib/dash/dash_parser.js [hunk 3]
```diff
@@ -528,25 +588,240 @@ shaka.dash.DashParser = class {
...
+
+  /**
+   * Handles common procedures after processing new periods.
+   *
+   * @param {!Array<shaka.extern.Period>} periods to be appended
+   * @param {boolean} isPatchUpdate does call comes from mpd patch update
+   * @private
+   */
+  async postPeriodProcessing_(periods, isPatchUpdate) {
+    await this.periodCombiner_.combinePeriods(periods, true, isPatchUpdate);
```
Comment (joeyparrish) [inline lib/dash/dash_parser.js:603]: This is another call to combinePeriods that wasn't in the original flow, executed even when `enablePatchMPDSupport` is false or `patchLocationUrl_` is `null`.

**Review by joeyparrish — CHANGES_REQUESTED** [review]
Overall, I would be comfortable with an "experimental" flag to enable this, but I want to ensure that the existing flow isn't broken.  I'm most concerned about extra calls to `combinedPeriods` (which can be expensive) and missing calls to set up caption streams.

File: lib/dash/dash_parser.js [hunk 4]
```diff
@@ -56,6 +56,26 @@ shaka.dash.DashParser = class {
...
+     */
+    this.manifestPatchContext_ = {
+      type: '',
+      profiles: [],
+      mediaPresentationDuration: null,
+      availabilityTimeOffset: 0,
+      baseUris: [],
+    };
+
+    /** @private {!Map<string, !shaka.dash.DashParser.Context>} */
```
[the remark this answers is not in the document]
Comment (dave-nicholas) [reply lib/dash/dash_parser.js]: done

File: lib/dash/dash_parser.js [hunk 5]
```diff
@@ -2051,7 +2416,9 @@ shaka.dash.DashParser.RequestSegmentCallback;
  *   language: ?string,
  *   numChannels: ?number,
  *   audioSamplingRate: ?number,
- *   availabilityTimeOffset: number
+ *   availabilityTimeOffset: number,
+ *   initialization: ?string,
```
[the remark this answers is not in the document]
Comment (dave-nicholas) [reply lib/dash/dash_parser.js:2470]: done

File: lib/dash/dash_parser.js [hunk 6]
```diff
@@ -2023,6 +2363,31 @@ shaka.dash.DashParser = class {
...
+
+
+/**
+ * @typedef {{
+ *   timescale: number,
+ *   duration: number,
+ *   startNumber: number,
+ *   presentationTimeOffset: number,
+ *   media: ?string,
+ *   index: ?string,
```
[the remark this answers is not in the document]
Comment (dave-nicholas) [reply lib/dash/dash_parser.js:2411]: done

File: lib/dash/dash_parser.js [hunk 7]
```diff
@@ -2023,6 +2363,31 @@ shaka.dash.DashParser = class {
   }
 };
 
+/**
+ * @typedef {{
+ *   type: string,
```
[the remark this answers is not in the document]
Comment (dave-nicholas) [reply lib/dash/dash_parser.js:2382]: done

File: lib/dash/dash_parser.js (hunk 1 above)
Comment (dave-nicholas) [reply lib/dash/dash_parser.js:562]: The call above is the one I should have removed but left in this PR, sorry.

File: lib/dash/dash_parser.js (hunk 2 above)
Comment (dave-nicholas) [reply lib/dash/dash_parser.js:549]: This has been reverted.

File: lib/dash/dash_parser.js (hunk 3 above)
Comment (dave-nicholas) [reply lib/dash/dash_parser.js:613]: This is the only call now during an update.

File: lib/dash/dash_parser.js [hunk 8]
```diff
@@ -1462,6 +1754,53 @@ shaka.dash.DashParser = class {
...
+          pixelAspectRatio: context[k].pixelAspectRatio,
+          emsgSchemeIdUris: context[k].emsgSchemeIdUris,
+          id: context[k].id,
+          numChannels: context[k].numChannels,
+          audioSamplingRate: context[k].audioSamplingRate,
+          availabilityTimeOffset: context[k].availabilityTimeOffset,
+          initialization: context[k].initialization,
+        };
+      } else if (k == 'periodInfo') {
+        contextClone[k] = {
```
[the remark this answers is not in the document]
Comment (dave-nicholas) [reply lib/dash/dash_parser.js:1804]: done

**Comment by avelad** [comment]
@dave-nicholas @tykus160 Please add a Demo streams. You can use https://github.com/Dash-Industry-Forum/dash.js/issues/4452 . Thanks!

**Comment by tobbee** [comment]
When running towards dash.js, some issues were fond in the ongoing development of MPD patch in livesim2. I think things should be good now, but if you find any issues, please report them to https://github.com/Dash-Industry-Forum/livesim2/pull/174 or open a new issue if the PR is closed.

**Comment by tobbee** [comment]
The MPD Patch support is now on the public DASH-IF livesim2 server. Here is a possible test URL:

https://livesim2.dashif.org/livesim2/patch_60/segtimeline_1/testpic_2s/Manifest.mpd

**Comment by tykus160** [comment]
Added demo asset, thanks @tobbee. This PR still needs some love though, as initial scope of work is not sufficient to properly play DASH-IF asset.

**Comment by tobbee** [comment]
@avelad The output is "live" streams and available in three different formats:

* SegmentTemplate with `$Number$`: https://livesim2.dashif.org/livesim2/patch_60/testpic_2s/Manifest.mpd
* SegmentTimeline with `$Number$`: https://livesim2.dashif.org/livesim2/patch_60/segtimelinenr_1/testpic_2s/Manifest.mpd
* SegmentTimeline with `$Time$`: https://livesim2.dashif.org/livesim2/patch_60/segtimeline_1/testpic_2s/Manifest.mpd

`SegmentList` is not a format recommended by DASH-IF so it is not supported.
`SegmentBase` is only useful for VoD content  in DASH OnDemand profile as far as I know, so it is irrelevant to live output.

I hope you can make these variants work (the first one has very little updates since no Segments are mentioned).

Once they work, one can also test with multiperiod variants by adding `/periods_60` before `testpic_2s` in any of the above URLs like

https://livesim2.dashif.org/livesim2/patch_60/segtimeline_1/periods_60/testpic_2s/Manifest.mpd

**Comment by tykus160** [comment]
Hi! This PR now supports DASH IF test streams, though there are few gotchas @tobbee 
- SegmentTemplate with `$Number$` has `publishTime` set to 1970, so we do regular updates there instead of patches
- on multiperiod, whenever new period arrives, following patch requests will return HTTP 500, something similar has been reported on dash.js board as well https://github.com/Dash-Industry-Forum/dash.js/issues/4487

**Review by avelad — APPROVED** [review]

**Comment by avelad** [comment]
@shaka-bot test

**Comment by tobbee** [comment]
@tykus160 The patch issues in livesim2 should be fixed now including the multiperiod case for SegmentTimeline.

Regarding the publishTime for SegmentTemplate with $Number$ it changes every time the MPD changes, which is every minute for  one-minute period, but never if there are no periods. The reason behind that is that it the publishTime is an "id" of the manifest so it shouldn't change every segment if there is no change.

In general, I think that MPD Patch and SegmentTemplate with Number is a bad match, so one should rather focus on SegmentTimeline, which is also the case with long MPDs. I think that your choice of regular updates is the most reasonable in that case.

As far as I understood from Daniel Silhavy, dash.js fallbacks to normal MPD requests if there is no update of MPD patch.

**Review by avelad — APPROVED** [review]

**Review by joeyparrish — APPROVED** [review]

**Pull request merged**